src/audio/volume.c

```diff
@@ -250,6 +257,52 @@ static inline void volume_set_chan(struct comp_dev *dev, int chan, uint32_t vol)
 		v = VOL_MAX;
 
 	cd->tvolume[chan] = v;
+
+	/* Check ramp type */
+	switch (pga->ramp) {
+	case SOF_VOLUME_LINEAR:
+
+		/* Assume the ramp length (initial_ramp [ms]) describes
+		 * time of mute to 0 dB ramp. The actual volume scale min/max
+		 * is not known. If initial_ramp is zero (or negative) use
+		 * step size that reaches 0 dB gain in one step (VOL_ZERO_DB).
+		 */
+		if (pga->initial_ramp > 0)
+			inc = MAX(VOL_RAMP_STEP_CONST / pga->initial_ramp, 1);
+		else
+			inc = VOL_ZERO_DB;
+
+		delta = cd->tvolume[chan] - cd->volume[chan];
+		delta_abs = ABS(delta);
+
+		/* If the ramp completion would take longer than initial_ramp
+		 * increase the step size to make this ramp complete at
+		 * initial_ramp time. Calculate estimated ramp length [us].
+		 */
+		t_us = delta_abs * VOL_RAMP_UPDATE_US / inc;
+		if (t_us > pga->initial_ramp * 1000) {
```

Comment:
@singalsu for my understanding when will this ever happen? the target volume is limited to VOL_MAX or VOL_MIN. so the delta can ever be maximum of VOL_ZERO_DB isnt it?

Reply:
@ranj063 This happens pretty often when volume control is above 0 dB. This step re-calculate is the beef of the slow gain settling fix (while kernel does not provide the used min/max range). 

VOL_MAX is +42 dB so the the linear mute->unity gain equal transition is only e.g. 0 dB -> +6 dB jump. Such volume transitions can be made by pressing keys 0 - 9 in alsamixer ASCII graphics view.